### PR TITLE
[SPARK-32559][SQL] Fix the trim logic did't handle ASCII control characters correctly

### PR DESCRIPTION
**Pull request by beryllw**
### What changes were proposed in this pull request?
The trim logic in Cast expression introduced in https://github.com/apache/spark/pull/29375 trim ASCII control characters unexpectly.

Before this patch
image
And hive
image

### Why are the changes needed?
The behavior described above doesn't consistent with the behavior of Hive

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
add ut

### CODE REVIEW

**Comment by Kwafoor** [comment]
@cloud-fan @yaooqinn @gengliangwang @WangGuangxin Could you please help review this?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -535,14 +535,14 @@ public UTF8String trim() {
   public UTF8String trimAll() {
     int s = 0;
     // skip all of the whitespaces in the left side
-    while (s < this.numBytes && Character.isWhitespace(getByte(s))) s++;
+    while (s < this.numBytes && getByte(s) > 0x00 && getByte(s) <= 0x20) s++;
```
[the remark this answers is not in the document]
Comment (Kwafoor) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:538]: You're right, I'll make some adjustments and test the behavior in Hive to see how it performs.

**Comment by Kwafoor** [comment]
> I'm not sure this is correct. SQL TRIM from other DBs seems to default to removing _whitespace_ or the space char, not control codes.

I conducted tests on both MySQL and Hive, and found that Hive supports trimming control characters and spaces but does not support trimming characters within the range 0x7F to 0x9F. MySQL, on the other hand, supports trimming specific control characters.
image

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala [hunk 2]
```diff
@@ -127,6 +127,10 @@ class DateTimeUtilsSuite extends SparkFunSuite with Matchers with SQLHelper {
     stringToDate(UTF8String.fromString(s))
   }
 
+  test("string to date trim Control Characters") {
```
Comment (yaooqinn) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala:130]:
```suggestion
  test("SPARK-32559: string to date trim Control Characters") {
```

**Review by yaooqinn — APPROVED** [review]
As Spark is a general engine supporting various kinds of data formats and sources, IMO, it's necessary to trim controls that are very likely to be involved during data exchange.

So, LGTM.

**Comment by yaooqinn** [comment]
Thanks, merged to master and 3.4/3.3

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @Kwafoor , @yaooqinn and all.
SPARK-32559 is delivered as Apache Spark 3.0.1 patch.
We need a new JIRA issue because this will land at Apache Spark 3.5.0/3.4.1/3.3.3 newly.

**Comment by yaooqinn** [comment]
@dongjoon-hyun Oops, My bad. For the current status, is linking this pull request to a new jira OK?

**Comment by Kwafoor** [comment]
> Hi, @Kwafoor , @yaooqinn and all. [SPARK-32559](https://issues.apache.org/jira/browse/SPARK-32559) is delivered as Apache Spark 3.0.1 patch. We need a new JIRA issue because this will land at Apache Spark 3.5.0/3.4.1/3.3.3 newly.

My mistake, should I raise a new issue to link this pr?

**Comment by yaooqinn** [comment]
@Kwafoor Yes

**Comment by Kwafoor** [comment]
> @Kwafoor Yes

I created an issue https://issues.apache.org/jira/browse/SPARK-44383, but how to link this pull request to a new jira？

I saw that spark 3.4.1 was released, with this pr and the wrong issue name SPARK-32559.I’m very sorry about this.

image

**Comment by yaooqinn** [comment]
I have manually resolved and re-targeted the links.

We probably  need to push a new PR to the spark-website repo to fix the release note @Kwafoor

**Comment by Kwafoor** [comment]
> I have manually resolved and re-targeted the links.
> 
> We probably need to push a new PR to the spark-website repo to fix the release note @Kwafoor

I created  [PR-466](https://github.com/apache/spark-website/pull/466) to the spark-website repo to fix the release note.

**Comment by yaooqinn** [comment]
thank you @Kwafoor

**Pull request closed without merging**